treeherder/perf/models.py

```diff
@@ -107,32 +108,94 @@ class PerformanceDatumManager(models.Manager):
-                time.sleep(sleep_time)
+        logger.warning('Cycling Perfherder data from {0} repository...'
+                       .format(repository.name))
+        max_timestamp = self._determine_max_timestamp(repository.name,
+                                                      cycle_interval,
+                                                      unsheriffed_repos_expire_days)
+
+        try:
+            self._delete_in_chunks(repository, max_timestamp, chunk_size, sleep_time, logger,
+                                   started_at, max_overall_runtime)
+        except NoDataCyclingAtAll as ex:
+            logger.warning('Exception: {}'.format(ex))
+            return
 
         # also remove any signatures which are (no longer) associated with
         # a job
+        logger.warning('Removing performance signatures with missing jobs...')
         for signature in PerformanceSignature.objects.filter(
                 repository=repository):
-            if not self.filter(signature=signature).exists():
+            if not self.filter(
+                    repository=repository,  # redundant, but leverages (repository, signature) compound index
+                    signature=signature).exists():
                 signature.delete()
 
+    def _delete_in_chunks(self, repository, max_timestamp, chunk_size, sleep_time, logger,
+                          started_at, max_overall_runtime):
+        from django.db import connection
+        any_succesful_attempt = False
+
+        with connection.cursor() as cursor:
+            while True:
+                if self._should_quit(started_at, max_overall_runtime):
+                    logger.info('Max runtime exceeded. Quitting cycling for repository {}'
+                                .format(repository.name))
+                    break
+
+                try:
+                    cursor.execute('''
+                        DELETE FROM `performance_datum`
+                        WHERE repository_id = %s AND
+                            push_timestamp < %s
+                        LIMIT %s
+                    ''', [repository.id, max_timestamp, chunk_size])
+                except Exception as ex:
+                    logger.warning('Failed to delete performance data chunk, while running "{}" query. '
+                                   .format(cursor._last_executed))
+
+                    if any_succesful_attempt is False:
+                        raise NoDataCyclingAtAll from ex
+                    break
+                else:
+                    deleted_rows = cursor.rowcount
+
+                    if deleted_rows == 0:
+                        break  # finished removing all expired data
+                    else:
+                        any_succesful_attempt = True
+                        logger.warning('Successfully deleted some rows')
+
+            if sleep_time:
+                # Allow some time for other queries to get through
+                time.sleep(sleep_time)
```

Comment:
There is no need to sleep.  Assuming each `execute()` is inside its own transaction:  Other work will be queued by the database while `execute()` runs.  When `execute()` is done, the other work will be run before the next `execute()`.